api/plgd/plgd_time.c

```diff
@@ -560,4 +598,243 @@ plgd_time_configure(bool use_in_mbedtls,
   g_oc_plgd_time.set_system_time_data = set_system_time_data;
 }
 
+#ifdef OC_CLIENT
+
+plgd_time_fetch_config_t
+plgd_time_fetch_config(const char *uri, const oc_endpoint_t *endpoint,
+                       bool disable_time_verification,
+                       int selected_identity_cred_id, uint16_t timeout,
+                       plgd_time_on_fetch_fn_t on_fetch, void *on_fetch_data)
+{
+  plgd_time_fetch_config_t fetch = {
+    .uri = uri,
+    .endpoint = endpoint,
+    .timeout = timeout,
+    .on_fetch = on_fetch,
+    .on_fetch_data = on_fetch_data,
+  };
+
+#if defined(OC_SECURITY) && defined(OC_PKI)
+  fetch.disable_time_verification = disable_time_verification;
+  fetch.selected_identity_cred_id = selected_identity_cred_id;
+#else  /* !OC_SECURITY || !OC_PKI */
+  (void)disable_time_verification;
+  (void)selected_identity_cred_id;
+#endif /* OC_SECURITY && OC_PKI */
+
+  return fetch;
+}
+
+static bool
+dev_time_parse_fetch_response(const oc_rep_t *rep, oc_clock_time_t *time)
+{
+  assert(clock != NULL);
+
+  const char *time_str = NULL;
+  size_t time_str_len = 0;
+  if (!oc_rep_get_string(rep, "time", (char **)&time_str, &time_str_len)) {
+    OC_ERR("fetch plgd-time failed: cannot find time property");
+    return false;
+  }
+
+  oc_clock_time_t ct = oc_clock_parse_time_rfc3339(time_str, time_str_len);
+  if (ct == (oc_clock_time_t)-1) {
+    OC_ERR("fetch plgd-time failed: parse clock time from string(%s)",
+           time_str);
+    return false;
+  }
+
+  *time = ct;
+  return true;
+}
+
+static void
+dev_time_on_fetch(oc_client_response_t *data)
+{
+  time_fetch_param_t *fp = (time_fetch_param_t *)data->user_data;
+  oc_clock_time_t time = 0;
+  oc_status_t code = data->code;
+  if (code == OC_STATUS_OK &&
+      !dev_time_parse_fetch_response(data->payload, &time)) {
+    code = OC_STATUS_INTERNAL_SERVER_ERROR;
+  }
+
+  OC_DBG("plgd-time: on_fetch time=%d time=%u", (int)code, (unsigned)time);
+  fp->on_fetch(code, time, fp->on_fetch_data);
+#if defined(OC_TCP) || defined(OC_SECURITY)
+  if (fp->close_peer_after_fetch) {
+    OC_DBG("plgd-time: close fetch time session");
+    oc_close_session(data->endpoint);
+  }
+#endif /* OC_TCP || OC_SECURITY */
+
+  oc_memb_free(&g_fetch_params_s, fp);
+}
+
+#if defined(OC_SECURITY) && defined(OC_PKI)
+
+static void
+time_verify_certificate_params_free(void *data)
+{
+  if (data == NULL) {
+    return;
+  }
+  time_verify_certificate_params_t *vcp =
+    (time_verify_certificate_params_t *)data;
+  oc_memb_free(&g_time_verify_certificate_params_s, vcp);
+}
+
+static int
+dev_time_verify_certificate(oc_tls_peer_t *peer, const mbedtls_x509_crt *crt,
+                            int depth, uint32_t *flags)
+{
+  OC_DBG("plgd-time: verifying certificate at depth %d, flags %u", depth,
+         *flags);
+
+  const time_verify_certificate_params_t *vcp =
+    (time_verify_certificate_params_t *)peer->user_data.data;
```

Comment:
set origin `user data` and `free` func to peer before this function is called.
`vcp->verify_certificate` can expect them